ARCs/arc-0031.md

```diff
@@ -0,0 +1,269 @@
+
+```mermaid
+sequenceDiagram
+    actor A as Alice
+    participant W as Wallet
+    actor B as Bob
+    A-->>W: Connect to Bob with PKa
+    activate W
+    W->>B: GET message for PKa
+    activate B
+    B->>B: Create a new msg for PKa
+    B->>W: msg
+    deactivate B
+    W->>W: Check <msg, Bob>
+    W-->>A: Show msg origin
+    Note right of A: Confirm signature
+    A-->>W: 
+    W->>B: <PKa, Sig(msg)>
+    deactivate W
+    activate B
+    B->>B: Verify Sig(msg) with PKa and msg
+    Note right of B: if Sig(msg) is valid<br/>then authenticate user
+    B-->>A: Authentication OK/KO
+    deactivate B
+```
+
+The diagram above summarizes the proposed mechanism. We consider the User, **Alice**, owner of an Algorand account <*PKa, SKa*> of which the secret key *SKa* is stored into a **Wallet**.
+
+> A wallet is any type of Algorand wallet, such as hot wallets like <a href="https://www.purestake.com/technology/algosigner/">AlgoSigner</a>, <a href="https://wallet.myalgo.com/">MyAlgo Wallet</a> for browser and mobile wallets used through <a href="https://developer.algorand.org/docs/get-details/walletconnect/">WalletConnect</a>, and cold wallets like the <a href="https://www.ledger.com">Ledger Nano</a>.
+
+Alice authenticates herself to the Verifier **Bob** sending back the digital signature `Sig(msg)` of message `msg` provided by Bob. The mechanism proceeds as follows:
+
+1. Alice initiates an authentication with *PKa* to Bob;
+2. Bob generates a message `msg` to be signed by Alice's *SKa*;
+3. Alice signs `msg` using her wallet; the Wallet inspects and displays the `msg` origin to be sure it came from Bob;
+4. Alice sends back the tuple `<PKa, Sig(msg)>` to Bob;
+5. Bob verifies `Sig(msg)` with Alice's *PKa* and `msg`;
+6. If the signature is valid, Bob authenticates Alice.
+
+The ARC-31 defines a standardized message for authentication, called *Authentication Message*.
+
+### Authentication Message
+
+An *Authentication Message* `auth-msg` is a sequence of bytes representing a message to be signed. The Verifier requests Users to sign an `auth-msg` with their secret keys. Such a message **MUST** include the following information:
+
+- `domain name`: name of the Verifier domain;
+- `Algorand address`: User's *PKa* to be authenticated;
+- `nonce`: a unique/random value generated by the Verifier;
+- `description`: Verifier details or general description;
+- `metadata`: arbitrary message data.
+
+The *Authentication Message* is represented with the following interface:
+
+```typescript
+interface AuthMessage {
+ /** The domain name of the Verifier */
+ domain: string;
+ /** Algorand account to authenticate with*/
+ authAcc: string;
+ /** Unique random nonce generated by the Verifier */
+ nonce: string;
```

Comment:
Just throwing a consideration here. Perhaps we should use **challenge** instead of nonce. 

There are similar auth standards being developed and proposed ([DIDAuth](https://w3c-ccg.github.io/vp-request-spec/#did-authentication), FIDO [U2F](https://web.archive.org/web/20230416040757/https://fidoalliance.org/specs/fido-u2f-v1.0-ps-20141009/fido-u2f-overview-ps-20141009.html) / [HID](https://web.archive.org/web/20220308142005/https://fidoalliance.org/specs/fido-u2f-v1.0-ps-20141009/fido-u2f-hid-protocol-ps-20141009.html))

Might be a small thing, but operating close to auth standards might be useful with integrations, complaince, etc.